### PR TITLE
Adding search queries for type and classes elements

### PR DESCRIPTION
**Pull request by RadioPotin**
Also: 
- some changes to `version.db`, like `CASCADE` and new tables for types, classes, and 2 additional ones for contents of types and classes

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -6,7 +6,7 @@ VERSION:=1.0
 .EXPORT_ALL_VARIABLES:
 PGDATABASE:=digodoc
 API_PORT:=11001
-DIGODOC_DIR:=/home/hernouf/DIGODOC/digodoc/_digodoc
+DIGODOC_DIR:=/home/dariocp/Programmation/OCPro/digodoc/_digodoc
```
Comment (zapashcanon) [inline Makefile:9]: this would be good to have a proper solution for this

File: src/db/misc_db.ml [hunk 2]
```diff
@@ -127,6 +127,50 @@ let val_of_row row opam_row mdl_row =
...
 
+let type_of_row_opt modules row opam_row mdl_row : Data_types.type_element option =
+    let opam_row = List.hd opam_row
+    and mdl_row = List.hd mdl_row in
+    let opampath = path_of_opam opam_row#opam_name opam_row#opam_version
+    and opam = name_of_opam opam_row#opam_name opam_row#opam_version
+    and mdlpath = mdl_row#mdl_path
+    and mdl = mdl_row#mdl_name in
+    (* If val resepcts one of opam and mdl condition *)
+    if List.length modules = 0
```
Comment (zapashcanon) [inline src/db/misc_db.ml:138]: `modules = []`

File: src/db/misc_db.ml [hunk 3]
```diff
@@ -127,6 +127,50 @@ let val_of_row row opam_row mdl_row =
...
+                mdlpath;
+                opam;
+                opampath
+            }
+    else None
+(** Creates [Data_types.type_element] from DB rows. *)
+
+let class_of_row_opt modules row opam_row mdl_row : Data_types.class_element option =
+    let opam_row = List.hd opam_row
+    and mdl_row = List.hd mdl_row in
```
Comment (zapashcanon) [inline src/db/misc_db.ml:154]: Same here about `and` and `List.hd`

File: src/db/misc_db.ml [hunk 4]
```diff
@@ -127,6 +127,50 @@ let val_of_row row opam_row mdl_row =
...
+    let opampath = path_of_opam opam_row#opam_name opam_row#opam_version
+    and opam = name_of_opam opam_row#opam_name opam_row#opam_version
+    and mdlpath = mdl_row#mdl_path
+    and mdl = mdl_row#mdl_name in
+    (* If val resepcts one of opam and mdl condition *)
+    if List.length modules = 0
```
Comment (zapashcanon) [inline src/db/misc_db.ml:160]: `modules = []`

File: src/db/versions.ml [hunk 5]
```diff
@@ -56,33 +56,56 @@ let init () =
...
         primary key (mdl_id, mdl_ident)
-      )|}
-    ]
+      )|};
+      {|create table module_types(
+        mdl_id int not null references module_index(mdl_id),
+        mdl_name varchar not null,
+        mdl_opam_name varchar not null references opam_index(opam_name),
+        type_id int not null primary key,
+        ident varchar not null
```
Comment (zapashcanon) [inline src/db/versions.ml:75]: could you use uppercase for SQL keywords ?

File: src/db/misc_db.ml (hunk 4 above)
Comment (RadioPotin) [reply src/db/misc_db.ml:160]: I don't think that can happen if you look at `src/db/db.ml` [line 163](https://github.com/OCamlPro/digodoc-search-api/blob/b5331865ed8543c5bd9759b32497836207e31cef/src/db/db.ml#L163), this is the function that return the value stored in variable `modules` [here](https://github.com/OCamlPro/digodoc-search-api/blob/b5331865ed8543c5bd9759b32497836207e31cef/src/api/handlers.ml#L62), it's value cannot be `[]`, hence the lack of matching on that variable, and the subsequent calls to `List.hd`

File: src/db/misc_db.ml [hunk 6]
```diff
@@ -127,6 +127,50 @@ let val_of_row row opam_row mdl_row =
     }
 (** Creates [Data_types.val_element] from DB rows. *)
 
+let type_of_row_opt modules row opam_row mdl_row : Data_types.type_element option =
+  let opam_row = List.hd opam_row in
```
Comment (zapashcanon) [inline src/db/misc_db.ml:131]: match on this properly

File: src/db/misc_db.ml [hunk 7]
```diff
@@ -127,6 +127,50 @@ let val_of_row row opam_row mdl_row =
...
+  let mdl_row = List.hd mdl_row in
```
Comment (zapashcanon) [inline src/db/misc_db.ml:132]: match on this properly

File: src/db/misc_db.ml [hunk 8]
```diff
@@ -127,6 +127,50 @@ let val_of_row row opam_row mdl_row =
...
+                mdl;
+                mdlpath;
+                opam;
+                opampath
+            }
+    else None
+(** Creates [Data_types.type_element] from DB rows. *)
+
+let class_of_row_opt modules row opam_row mdl_row : Data_types.class_element option =
+  let opam_row = List.hd opam_row in
```
Comment (zapashcanon) [inline src/db/misc_db.ml:153]: match on this properly

File: Makefile [hunk 9]
```diff
@@ -6,7 +6,7 @@ VERSION:=1.0
 .EXPORT_ALL_VARIABLES:
 PGDATABASE:=digodoc
 API_PORT:=11001
-DIGODOC_DIR:=/home/hernouf/DIGODOC/digodoc/_digodoc
+DIGODOC_DIR ?=/home/dariocp/Programmation/OCPro/digodoc/_digodoc
```
Comment (zapashcanon) [inline Makefile:9]: it would be good to have a proper solution for this

**Pull request merged**